src/classes/RD_RecurringDonations.cls

```diff
@@ -795,77 +766,391 @@ public class RD_RecurringDonations {
+
+            if (needsRecalc) {
+                if(r.npe03__Open_Ended_Status__c == closedlabel
+                        && oldRecurringDonations.get(r.id).npe03__Open_Ended_Status__c != r.npe03__Open_Ended_Status__c
                     ){
-                    reevaluateopps.add(r.id);
+                    // closing an open ended donation
+                    // does it actually matter if it's newly closed? shouldn't it go to the runRecurringDonationClosedActions method regardless?
+                    // TODO: confirm this
+                    newlyClosedDonations.add(r);
+                } else if (r.npe03__Open_Ended_Status__c == openlabel) {
+                    // open-ended
+                    openEndedRDsToRecalculate.add(r);
+                } else {
+                    // fixed-length
+                    fixedLengthRDsToRecalculate.add(r);
                 }
-                else if (r.npe03__Installment_Period__c != oldRD.npe03__Installment_Period__c || r.npe03__Contact__c != oldRD.npe03__Contact__c){
-                    reinsertopps.add(r);
-                }                                        
             }
-        }      
+        }
         
-        if (!newlycloseddonations.isEmpty()){
-            runRecurringDonationClosedActions(newlycloseddonations, dmlWrapper);
+        if (!newlyClosedDonations.isEmpty()){
+            runRecurringDonationClosedActions(newlyClosedDonations, dmlWrapper);
         }
-        if (!reevaluateopps.isEmpty()){
-            if (reevaluateopps.size() == 1 || system.isFuture() || system.isBatch())
-                updateExistingOpps(reevaluateopps, dmlWrapper);
-            else
-                updateExistingOppsFuture(reevaluateopps);
+
+        if (!openEndedRDsToRecalculate.isEmpty()) {
+            List<Opportunity> openOpps = [SELECT Id
+                    , Amount
+                    , Name
+                    , CampaignId
+                    , AccountID
+                    , CloseDate
+                    , StageName
+                    , Recurring_Donation_Installment_Number__c
+                    , npe03__Recurring_Donation__c
+            FROM Opportunity
+            WHERE npe03__Recurring_Donation__c IN :openEndedRDsToRecalculate
+            AND isClosed = false
+            ORDER BY CloseDate ASC];
+            handleOpenEndedScheduleRecalculation(openEndedRDsToRecalculate, openOpps, dmlWrapper);
         }
-        //needs a total reinsert, not an update
-        if (!reinsertopps.isEmpty()){
-            //clear existing open opps and recreate
-            delete [select id from Opportunity where npe03__Recurring_Donation__c IN :reinsertopps and isClosed = false];
-            if (reinsertopps.size() == 1 || system.isFuture() || system.isBatch()){
-                insertOppsOnRecurringDonationInsert(reinsertopps);                  
+
+        if (!fixedLengthRDsToRecalculate.isEmpty()) {
+            List<Opportunity> opps = [SELECT Id
```

Comment:
I'm pretty sure this code gets called in a trigger context, so ideally it should limit the number of SOQL queries. These 2 queries are nearly identical except for the filter. It might be better to combine into a single query and then split the data in a for loop into two collections for passing to the appropriate method for handling.